### ISSUE TITLE
Inference for single row arff file fails

### ISSUE THREAD

**Issue by PGijsbers** [h2oai/h2o-3#15572]
**H2O version, Operating System and Environment**
Ran in docker with host `M1 MacOS` and container `ubuntu:22.04`. 
Python `3.9.17` and `h2o=='3.40.0.4'`
```
Java Version: openjdk version "1.8.0_362"; OpenJDK Runtime Environment (build 1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09); OpenJDK 64-Bit Server VM (build 25.362-b09, mixed mode)
```

**Actual behavior**
I expect h2o to be able to produce prediction for single row arff files.

**Expected behavior**
Instead, an error is shown that there are "no columns in common" (stacktrace below).
However, the arff headers are identical and the data is valid in both, so I do not understand where the error comes from.
Both files also end with a newline.

**Steps to reproduce**
I tried to set up a MWE, but I did run it from within the [automlbenchmark](www.github.com/openml/automlbenchmark)) since I did not want to set up an H2O server on my host machine.

```
import h2o
from h2o.automl import H2OAutoML
import subprocess

# Start the H2O cluster (locally)
h2o.init()

# Download Iris dataset
subprocess.call(["wget", "https://www.openml.org/data/download/61/dataset_61_iris.arff"])
# Create a subsample of just one row
with open("dataset_61_iris.arff", "r") as input_:
    arff = input_.readlines()

data_start = arff.index("@DATA\n")
subsample = arff[:data_start+2]
print(subsample[-3:])

with open("iris_subsample.arff", "w") as output:
    output.write(''.join(subsample))

train = h2o.import_file("dataset_61_iris.arff")
test = h2o.import_file("iris_subsample.arff")

aml = H2OAutoML(max_models=1, seed=1)
aml.train(y="class", training_frame=train)
aml.predict(test)  # raises the error
```

**Upload logs**
I couldn't find them (likely due to the AMLB setup).

**Screenshots**
N/A

**Additional context**
Stacktrace (I ran the snippet from within the [automlbenchmark](www.github.com/openml/automlbenchmark)).
```
ERROR:frameworks.shared.callee:Job with key $03017f000001120effffffff$_8d83ecf1772d78754e34f88ec4b14f6e failed with an exception: java.lang.IllegalArgumentException: Test/Validation dataset has no columns in common with the training set

stacktrace: 
2023-06-13 21:17:07 (5.34 MB/s) - ‘dataset_61_iris.arff.2’ saved [7487/7487]
java.lang.IllegalArgumentException: Test/Validation dataset has no columns in common with the training set

	at hex.Model.adaptTestForTrain(Model.java:1790)

	at hex.Model.adaptTestForTrain(Model.java:1598)

	at hex.Model.adaptTestForTrain(Model.java:1594)

	at hex.Model.adaptFrameForScore(Model.java:1925)

	at hex.Model.score(Model.java:1943)

	at water.api.ModelMetricsHandler$1.compute2(ModelMetricsHandler.java:516)

	at water.H2O$H2OCountedCompleter.compute(H2O.java:1677)

	at jsr166y.CountedCompleter.exec(CountedCompleter.java:468)

	at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263)

	at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:976)

	at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1479)

	at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)

Traceback (most recent call last):

  File "/bench/frameworks/shared/callee.py", line 74, in call_run

    result = run_fn(ds, config)

  File "/bench/frameworks/H2OAutoML/exec.py", line 95, in run

    aml.predict(test)

  File "/bench/frameworks/H2OAutoML/venv/lib/python3.9/site-packages/h2o/automl/_estimator.py", line 695, in predict

    return leader.predict(test_data)

  File "/bench/frameworks/H2OAutoML/venv/lib/python3.9/site-packages/h2o/model/model_base.py", line 323, in predict

    j.poll()

  File "/bench/frameworks/H2OAutoML/venv/lib/python3.9/site-packages/h2o/job.py", line 90, in poll

    raise EnvironmentError("Job with key {} failed with an exception: {}\nstacktrace: "

OSError: Job with key $03017f000001120effffffff$_8d83ecf1772d78754e34f88ec4b14f6e failed with an exception: java.lang.IllegalArgumentException: Test/Validation dataset has no columns in common with the training set

stacktrace: 

java.lang.IllegalArgumentException: Test/Validation dataset has no columns in common with the training set

	at hex.Model.adaptTestForTrain(Model.java:1790)

	at hex.Model.adaptTestForTrain(Model.java:1598)

	at hex.Model.adaptTestForTrain(Model.java:1594)

	at hex.Model.adaptFrameForScore(Model.java:1925)

	at hex.Model.score(Model.java:1943)

	at water.api.ModelMetricsHandler$1.compute2(ModelMetricsHandler.java:516)

	at water.H2O$H2OCountedCompleter.compute(H2O.java:1677)

	at jsr166y.CountedCompleter.exec(CountedCompleter.java:468)

	at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263)

	at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:976)

	at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1479)

	at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)






Job with key $03017f000001120effffffff$_8d83ecf1772d78754e34f88ec4b14f6e failed with an exception: java.lang.IllegalArgumentException: Test/Validation dataset has no columns in common with the training set
stacktrace: 
java.lang.IllegalArgumentException: Test/Validation dataset has no columns in common with the training set
	at hex.Model.adaptTestForTrain(Model.java:1790)
	at hex.Model.adaptTestForTrain(Model.java:1598)
	at hex.Model.adaptTestForTrain(Model.java:1594)
	at hex.Model.adaptFrameForScore(Model.java:1925)
	at hex.Model.score(Model.java:1943)
	at water.api.ModelMetricsHandler$1.compute2(ModelMetricsHandler.java:516)
	at water.H2O$H2OCountedCompleter.compute(H2O.java:1677)
	at jsr166y.CountedCompleter.exec(CountedCompleter.java:468)
	at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263)
	at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:976)
	at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1479)
	at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)
Traceback (most recent call last):
  File "/bench/amlb/benchmark.py", line 563, in run
    meta_result = self.benchmark.framework_module.run(self._dataset, task_config)
  File "/bench/frameworks/H2OAutoML/__init__.py", line 30, in run
    return run_in_venv(__file__, "exec.py",
  File "/bench/frameworks/shared/caller.py", line 152, in run_in_venv
    raise NoResultError(res.error_message)
amlb.results.NoResultError: Job with key $03017f000001120effffffff$_8d83ecf1772d78754e34f88ec4b14f6e failed with an exception: java.lang.IllegalArgumentException: Test/Validation dataset has no columns in common with the training set
stacktrace: 
java.lang.IllegalArgumentException: Test/Validation dataset has no columns in common with the training set
	at hex.Model.adaptTestForTrain(Model.java:1790)
	at hex.Model.adaptTestForTrain(Model.java:1598)
	at hex.Model.adaptTestForTrain(Model.java:1594)
	at hex.Model.adaptFrameForScore(Model.java:1925)
	at hex.Model.score(Model.java:1943)
	at water.api.ModelMetricsHandler$1.compute2(ModelMetricsHandler.java:516)
	at water.H2O$H2OCountedCompleter.compute(H2O.java:1677)
	at jsr166y.CountedCompleter.exec(CountedCompleter.java:468)
	at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263)
	at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:976)
	at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1479)
	at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)
```

**Comment by tomasfryda** [comment]
Thank you for reporting this issue @PGijsbers. I believe there is a bug that checks the number of "data" lines and if there is just one it assumes there is no header -> the test set has automatically generated column names `C1, C2,...` and they don't match the names from the training set.

I don't know when we'll get to properly fix this but as a workaround you can use something like `test = h2o.import_file("iris_subsample.arff", col_names=train.names)` or add another line (it can even be just missing values(in this case it would be `,,,,\n`) and then subset to pick just the first line.

**Comment by tomasfryda** [comment]
@PGijsbers This is fixed in the upcoming release (3.42.0.1) that should be out soon (end of this week or the next week).
